modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManagerSelfTest.java

```diff
@@ -0,0 +1,901 @@
+            (part, grp) -> {});
+
+        fut2.get();
+    }
+
+    /**
+     * @throws Exception If fails.
+     */
+    @Test
+    public void testSnapshotRemoteOnBothNodes() throws Exception {
+        IgniteEx ig0 = startGrids(2);
+
+        ig0.cluster().active(true);
+
+        for (int i = 0; i < CACHE_KEYS_RANGE; i++)
+            ig0.cache(DEFAULT_CACHE_NAME).put(i, i);
+
+        CheckpointProgress cpFut = ig0.context()
+            .cache()
+            .context()
+            .database()
+            .forceCheckpoint("the next one");
+
+        cpFut.futureFor(CheckpointState.FINISHED).get();
+
+        IgniteSnapshotManager mgr0 = ig0.context()
+            .cache()
+            .context()
+            .snapshotMgr();
+
+        IgniteSnapshotManager mgr1 = grid(1).context()
+            .cache()
+            .context()
+            .snapshotMgr();
+
+        UUID node0 = grid(0).localNode().id();
+        UUID node1 = grid(1).localNode().id();
+
+        // Snapshot must be taken on node1 and transmitted to node0.
+        IgniteInternalFuture<?> futFrom1To0 = mgr0.createRemoteSnapshot(node1,
+            owningParts(ig0, new HashSet<>(Collections.singletonList(CU.cacheId(DEFAULT_CACHE_NAME))), node1),
+            (part, grp) -> {});
+
+        IgniteInternalFuture<?> futFrom0To1 = mgr1.createRemoteSnapshot(node0,
+            owningParts(grid(1), new HashSet<>(Collections.singletonList(CU.cacheId(DEFAULT_CACHE_NAME))), node0),
+            (part, grp) -> {});
+
+        futFrom0To1.get();
+        futFrom1To0.get();
+    }
+
+    /**
+     * @throws Exception If fails.
+     */
+    @Test(expected = ClusterTopologyCheckedException.class)
+    public void testRemoteSnapshotRequestedNodeLeft() throws Exception {
+        IgniteEx ig0 = startGridWithCache(defaultCacheCfg, CACHE_KEYS_RANGE);
+        IgniteEx ig1 = startGrid(1);
+
+        ig0.cluster().setBaselineTopology(ig0.cluster().forServers().nodes());
+
+        awaitPartitionMapExchange();
+
+        CountDownLatch hold = new CountDownLatch(1);
+
+        ((GridCacheDatabaseSharedManager)ig1.context()
+            .cache()
+            .context()
+            .database())
+            .waitForCheckpoint("Snapshot before request", f -> {
+                try {
+                    // Listener will be exectuted inside the checkpoint thead.
+                    U.await(hold);
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException(e);
+                }
+            });
+
+        UUID rmtNodeId = ig1.localNode().id();
+
+        ig0.context()
+            .cache()
+            .context()
+            .snapshotMgr()
+            .createRemoteSnapshot(rmtNodeId,
+            owningParts(ig0,
+                new HashSet<>(Collections.singletonList(CU.cacheId(DEFAULT_CACHE_NAME))),
+                rmtNodeId),
+                (part, grp) -> {});
```

Comment:
I think no, `null` for can be used only in tests and never in a production environment since it has no meaning creating snapshots without any consumers.